### DISCUSSION TITLE
Are there build instructions? Yarn command fails.

### DISCUSSION THREAD

**Discussion by yurivict** [concourse/concourse#7118] [Help & Support]
I couldn't find build instructions in the documentation: https://concourse-ci.org/install.html The "Install" section doesn't contain instructions how to build it.

-------------------------

From the Arch port build procedure I concluded that the first command is probably "yarn" (https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=concourse).

But the ```yarn``` command fails:
```
[3/4] Linking dependencies...
[4/4] Building fresh packages...
[1/5] ⠠ elm
[-/5] ⠠ waiting...
[3/5] ⠠ elm-format
[4/5] ⠠ elmi-to-json
error /usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer: Command failed.
Exit code: 1
Command: node install.js
Arguments: 
Directory: /usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer
Output:
/usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer/lib/helper.js:279
    throw new Error(message);
    ^

Error: Unsupported platform: freebsd
    at assert (/usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer/lib/helper.js:279:11)
    at new BrowserFetcher (/usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer/lib/BrowserFetcher.js:96:7)
    at module.exports.createBrowserFetcher (/usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer/lib/Puppeteer.js:82:12)
    at Object.<anonymous> (/usr/ports/misc/concourse/work/concourse-7.1.0/node_modules/puppeteer/install.js:39:34)
    at Module._compile (node:internal/modules/cjs/loader:1092:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1121:10)
    at Module.load (node:internal/modules/cjs/loader:972:32)
```

-----------------------

Another concern is that NodeJS is a security risk - it downloads packages straight from GitHub and runs them immediately.

Could you please make pre-built web frontend available as a separate tarball? The Authelia project does this: 
https://github.com/authelia/authelia/releases, for example.

**Comment by taylorsilva** [comment]
> From the Arch port build procedure I concluded that the first command is probably "yarn" (https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=concourse).

Didn't know about this. We don't maintain (or even know about) that port. None of the people listed as users/maintainers on that are familiar to me in the Github or discord community. 

You can find our pre-built web and worker tarballs on the release page: https://github.com/concourse/concourse/releases

**Comment by yurivict** [comment]
> You can find our pre-built web and worker tarballs on the release page: https://github.com/concourse/concourse/releases

I'm only seeing pre-built binaries at this URL. Where are the web tarballs?

-----------------------

Another question is: What is the minimal set of node modules required to run Concourse? What causes concern is the "puppeteer``` module that fails on FreeBSD (see error in the OP), Puppeteer downloads binaries of various browsers (see here https://github.com/puppeteer/puppeteer/blob/main/experimental/puppeteer-firefox/lib/BrowserFetcher.js)

Downloading binaries is problematic on many levels:
1. Downloaded binaries might be compromised. Their origin, who compiled them is unclear.
2. We already have all browsers installed from packages.
3. Why having externally built browsers is required to run Concourse?

Concourse build is failing because browser binaries aren't available for FreeBSD. Are they really required?

**Reply by taylorsilva** [reply]
> I'm only seeing pre-built binaries at this URL. Where are the web tarballs?

Ah I think I see what you were expecting. All the html, js, css files are packed into the `concourse` binary, they are not separate from the binary, at least not the way we currently ship. When I said "web and worker tarballs" I was talking about the tarballs for each platform (linux, windows, darwin). The web and worker code is all in the same binary.

> What is the minimal set of node modules required to run Concourse?
> Concourse build is failing because browser binaries aren't available for FreeBSD. Are they really required?

@aoldershaw or @chenbh can probably better answer that.
I don't think you need the puppeteer module/browser binaries, I think that's only used for running tests.

**Reply by aoldershaw** [reply]
Yeah, puppeteer is only used for running the [`watsjs`](https://github.com/concourse/concourse/tree/master/web/wats) test suite. It does seem odd that it's in the main `package.json` as well - I'll try to remove it and see what breaks

**Reply by aoldershaw** [reply]
Removed in https://github.com/concourse/concourse/pull/7119

**Reply by aoldershaw** [reply]
...actually, it looks like it's also used for `yarn benchmark`. That's annoying, wonder if there's any way around that

**Reply by yurivict** [reply]
Another binary-downloading module is ```binwrap```:
```
error /disk-samsung/freebsd-ports/misc/concourse/work/github.com/concourse/concourse@v1.6.1-0.20210602163829-f1b2a420093f/node_modules/elm-format: Command failed.
Exit code: 1
Command: binwrap-install
Arguments: 
Directory: /disk-samsung/freebsd-ports/misc/concourse/work/github.com/concourse/concourse@v1.6.1-0.20210602163829-f1b2a420093f/node_modules/elm-format
Output:
/disk-samsung/freebsd-ports/misc/concourse/work/github.com/concourse/concourse@v1.6.1-0.20210602163829-f1b2a420093f/node_modules/binwrap/install.js:18
    throw new Error("No binaries are available for your platform: " + buildId);
    ^

Error: No binaries are available for your platform: freebsd-x64
    at install (/disk-samsung/freebsd-ports/misc/concourse/work/github.com/concourse/concourse@v1.6.1-0.20210602163829-f1b2a420093f/node_modules/binwrap/install.js:18:11)
    at Object.install (/disk-samsung/freebsd-ports/misc/concourse/work/github.com/concourse/concourse@v1.6.1-0.20210602163829-f1b2a420093f/node_modules/binwrap/index.js:14:14)
    at Object.<anonymous> (/disk-samsung/freebsd-ports/misc/concourse/work/github.com/concourse/concourse@v1.6.1-0.20210602163829-f1b2a420093f/node_modules/binwrap/bin/binwrap-install:18:9)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
```

Are binaries really needed for a server-side front-end?

**Reply by taylorsilva** [reply]
Back to your original question: Could you please make pre-built web frontend available as a separate tarball

I think this would be everything under the `web/public` folder. If you had that then you could skip entire workflow that your struggling to get through. Right?

**Comment by taylorsilva** [comment]
Going to convert in a discussion since there's no bug here. I'll reply in the discussion.

**Comment by yurivict** [comment]
> I think this would be everything under the web/public folder. If you had that then you could skip entire workflow that your struggling to get through. Right?

I think so.

For the web front-end you probably don't even need to run NodeJS build steps.  All they do - build or download binaries and compress JavaScripts.  You should only need JavaScripts. Binaries shouldn't be needed, and JavaScripts are usually available in pre-built form.

**Comment by yurivict** [comment]
Any chance to get pre-built web frontend tarballs (web/public folder) to the ```Releases``` page?

**Reply by taylorsilva** [reply]
Hey, this isn't on our roadmap but if someone from the community wants to help we accept PRs to all of our repos. The change would have to happen in our release pipeline template which is here: https://github.com/concourse/ci/blob/master/pipelines/release.yml

The `build-rc` job is probably a decent place to pull out the contents of the `web/public` folder and tar it up. We'd have to store this somewhere, the `concourse-artifacts` bucket would work: https://github.com/concourse/ci/blob/6e225e5a082e706d5cc4e746ec9c10b4a6d6b315/pipelines/release.yml#L1608-L1614

This can then be passed along to the `publish-binaries` job which publishes the github release: https://github.com/concourse/ci/blob/6e225e5a082e706d5cc4e746ec9c10b4a6d6b315/pipelines/release.yml#L1234-L1252

**Comment by yurivict** [comment]
This would be too much for me because I don't run concourse pipelines.

I am only trying to fix the FreeBSD port.
